src/modules/content-s3/s3.c

```diff
@@ -0,0 +1,268 @@
+
+    if (ctx.status != S3StatusOK && ctx.status != S3StatusErrorBucketAlreadyOwnedByYou) {
+        errno = EREMOTEIO;
+        if (errstr)
+            *errstr = S3_get_status_name (status);
+
+        return -1;
+    }
+
+    return 0;
+}
+
+int s3_put (struct s3_config *cfg, const char *key, const void *data, size_t size, const char **errstr)
+{
+    int retries = cfg->retries;
+    S3Status status = S3StatusOK;
+    
+    S3ResponseHandler resp_hndl = {
+        .propertiesCallback = &response_props_cb,
+        .completeCallback = &response_complete_cb
+    };
+
+    S3BucketContext bucket_ctx = {
+        0,
+        cfg->bucket,
+        protocol,
+        uri_style,
+        cfg->access_key,
+        cfg->secret_key
+    };
+
+    S3PutObjectHandler put_obj_hndl ={
+        .responseHandler = resp_hndl,
+        .putObjectDataCallback = &put_object_cb
+    };
+
+    if (strlen (key) == 0 || strchr (key, '/') || !strcmp (key, "..") || !strcmp (key, ".")) {
+        errno = EINVAL;
+        if (errstr)
+            *errstr = "invalid key";
+
+        return -1;
+    }
+
+    struct cb_data ctx = {
+        size,
+        (void *) data,
+        0,
+        status
+
+    };
+
+    do {
+        S3_put_object (&bucket_ctx, key, size, 0, 0, &put_obj_hndl, &ctx);
+        retries--;
+    } while (S3_status_is_retryable (ctx.status) && retries > 0);
+
+    if (ctx.status != S3StatusOK) {
+        errno = EREMOTEIO;
+        if (errstr)
+            *errstr = S3_get_status_name (ctx.status);
+
+        return -1;
+    }
+
+    return 0;
+}
+
+int s3_get (struct s3_config *cfg, const char *key, void **datap, size_t *sizep, const char **errstr)
+{
+    int retries = cfg->retries;
+    size_t size = 0;
+    void *data = NULL;
+    S3Status status = S3StatusOK;
+
+    S3ResponseHandler resp_hndl = {
+        .propertiesCallback = &response_props_cb,
+        .completeCallback = &response_complete_cb
+    };
+
+    S3BucketContext bucket_ctx = {
+        0,
+        cfg->bucket,
+        protocol,
+        uri_style,
+        cfg->access_key,
+        cfg->secret_key
+    };
+
+    uint64_t start = 0;
+    uint64_t len_bytes = 0;
+
+    S3GetObjectHandler get_obj_hndl = {
+        .responseHandler = resp_hndl,
+        .getObjectDataCallback =  &get_object_cb
+    };
+
+    struct cb_data ctx = {
+        0,
+        data,
+        size,
+        status
+    };
+   
+    if (strlen (key) == 0 || strchr (key, '/') || !strcmp (key, "..") || !strcmp (key, ".")) {
+        errno = EINVAL;
+        if (errstr)
+            *errstr = "invalid key";
+
+        return -1;
+    }
+
+    do {
+        S3_get_object (&bucket_ctx, key, 0, start, len_bytes, 0, &get_obj_hndl, &ctx);
+        retries--;
+    } while (S3_status_is_retryable (ctx.status) && retries > 0);
+
+    *datap = ctx.data;
+    *sizep = ctx.size;
+
+    free (data);
+
+    if (ctx.status != S3StatusOK) {
+        if (ctx.status == S3StatusErrorNoSuchKey)
```

Comment:
Since ctx.data could have been allocated before an error occured, it should be freed on the error path.  Best to do it before assigning errno since `free()` _could_ clobber it.